#12: avoid JS, #13: use parent, #14: increase UX

hohwille opened the pull request:
* #12: avoid JavaScript as Nashorn will be removed from future JDKs
* #13: use new `maven-parent` POM
* #14: increase UX
* changed artifactId from `devon-docgen` to `devonfw-docgen` to follow our standards.
* introcuded ci-friendly-maven

Please note that the `init` module is just a workaround to avoid redundancies.
Should we change the version of `init` to something like `dev-SNAPSHOT` to avoid staging it to maven central? I think this is a good idea - WDYT?

hohwille commented:
No feedback? Shall I merge and release and when someting is wrong I create a bugfix release?
I want to go ahead and make progress...

jambulud commented:
I'm checking the generation of devonfw-website. I will be able to say something once I finish the checks.

jambulud commented:
I have downloaded the repsitory and tried running "mvn clean package" in the root folder of the branch "feature/refactoring" but I get this error:

![image](https://user-images.githubusercontent.com/43032659/70795336-e491fe00-1d9f-11ea-8c87-47371b18bfd3.png)

hohwille commented:
You are using Windows that has a strange concept of file locking. When you have the PDF open in a PDF viewer you can not delete it on the disc unless you are closing the PDF viewer.
This is a general problem of windows and has nothing to do with docgen or asciidoctor at all. Please retry.

hohwille commented:
In case you do not know which process is locking your SVG and the errors does not go away then use `Process Explorer` to find the lock or even install `File Unlocker` (see devonfw-ide advanced tooling documentation).

jdiazgon commented:
I'm facing the same issue as @jambulud. 

```xml
[ERROR] Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:1.6.0:process-asciidoc (generate-doc) on project devonfw-docgen: Execution generate-doc of goal org.asciidoctor:asciidoctor-maven-plugin:1.6.0:process-asciidoc failed: org.jruby.exceptions.SystemCallError: (EACCES) Permission denied - C:/Users/jdiazgon/AppData/Local/Temp/image-20191216-27236-4y88c5.svg?label=license -> [Help 1]
```
> You are using Windows that has a strange concept of file locking

This is true. However, it seems that in the code there is a bug that we should fix.

hohwille commented:
FYI: I found a general problem with the changes due to ci-friendly-maven in combination with flatten-maven-plugin and our POM distribution. I fixed that and updated this PR to make it work.

hohwille commented:
> I'm facing the same issue as @jambulud.

As it seems the error is reproducible (at least on Windows). I am not sure why the error occurrs and would need to analyze and debug this in a Windows VM what requires some time that I currently do not have.

I assume that the error is more related to some version update I did implicitly rather than my actual indended changes. However, I fully agree we need to get this fixed, before we can merge this.

hohwille commented:
What I could find regarding versions:
I seem to use this version:
`<jruby.version>9.2.8.0</jruby.version>`
while asciidoctor (https://repo1.maven.org/maven2/org/asciidoctor/asciidoctor-maven-plugin/1.6.0/asciidoctor-maven-plugin-1.6.0.pom) seems to use this one:
`<jruby.version>9.2.6.0</jruby.version>`

So could you try to downgrade accordingly (or remove jruby dependency entirely) and see if that has influence on the error?

jambulud commented:
> So could you try to downgrade accordingly (or remove jruby dependency entirely) and see if that has influence on the error?

Mmm still not working after the downgrade in the root pom. Removing the dependency does not work either.

hohwille commented:
Actually I think the error has nothing to do with file-locks nor with asciidoctor versions.
The problem is that this asciidoc file is rendered during the build:
https://raw.githubusercontent.com/devonfw/docgen/91e164653dba9ffd477e8997ffe6d874f0e64a34/README.asciidoc
It contains this line:
> image:https://img.shields.io/maven-central/v/com.devonfw.tools/devon-docgen.svg?label=Maven%20Central["Maven Central",link=https://search.maven.org/search?q=g:com.devonfw.tools+a:devon-docgen]

Now asciidoctor seems to download this image to temporary directory resulting in something like
> C:/Users/jdiazgon/AppData/Local/Temp/image-20191216-27236-4y88c5.svg?label=license

Windows is not able to create a file containing this suffix `?label=license`.

hohwille commented:
So it is nice to have a README with all in one place. However, the esiest fix would be:
* create a `documentation/usage.asciidoc` file and move everything from `DocGen is a tool ...` on to that file (so we are devonfw documentation standard compliant)
* Remove that content accordingly from README and link to the usage file.

hohwille commented:
Should IMHO be fixed now and work as expected. Could you check again?
Please consider migration.asciidoc guide (`docgen.wiki.page` is now `docgen.asciidoc.page`).

hohwille commented:
I case you confirmed that JRuby version also is an issue please let me know, and I will update this PR asap.

jambulud commented:
> I case you confirmed that JRuby version also is an issue please let me know, and I will update this PR asap. Then also let me know if I you tested by taking jruby dependency out (or just changed the version variable).

I downgraded jruby but it is still not working

![image](https://user-images.githubusercontent.com/43032659/70996730-ae23ee00-20d3-11ea-9d63-c4d6a6077762.png)

I have even added https://github.com/devonfw/docgen/blob/master/pom.xml#L499 and https://github.com/devonfw/docgen/blob/master/pom.xml#L496 in html/pom.xml and docgen/pom.xml respectively but the error remains

jambulud commented:
I moved <eruby>erb</eruby> here https://github.com/devonfw/docgen/blob/feature/refactoring/pom.xml#L281, I'll post the results

hohwille commented:
Can you also provide the `pom.xml` of `devonfw-guide` (so we can check everything is correct there)?

jambulud commented:
devonfw-guide pom: https://github.com/devonfw/devonfw-guide/blob/master/pom.xml

devonfw website pom: https://github.com/devonfw/devonfw.github.io/blob/develop/pom.xml

hohwille commented:
I was asking for your modified pom. I assume you are missing out some required changes. Did you follow the migration hints?
I have locally changed `devonfw-guide` to use `5.0.0-SNAPSHOT` and can build the website as HTML.

jambulud commented:
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.devonfw.website</groupId>
  <artifactId>parent</artifactId>
  <version>dev-SNAPSHOT</version>
  <packaging>pom</packaging>

  <parent>
    <groupId>com.devonfw.tools</groupId>
    <artifactId>devonfw-docgen-html</artifactId>
    <version>5.0.0-SNAPSHOT</version>
  </parent>

  <modules>
    <module>asciidoctor-stylesheet</module>
    <module>website</module>
    <module>search-engine</module>
  </modules>

  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <docgen.images.dir>images</docgen.images.dir>
    <asciidoctor.templateDir>${basedir}/asciidoctor-backend/erb/html5/</asciidoctor.templateDir>
    <docgen.asciidoc.page>devonfw-guide/master</docgen.asciidoc.page>
    <stylesheet.target.dir>${basedir}/asciidoctor-stylesheet/target/stylesheets/</stylesheet.target.dir>
    <!-- The following includes the multipage html5 converter for #14, which does not work at the moment -->
    <!-- <asciidoctor.requires>${basedir}/asciidoctor-extensions/multipage-html5-converter.rb</asciidoctor.requires> -->
    <!-- <asciidoctor.maven.plugin.version>1.5.7.1</asciidoctor.maven.plugin.version> -->
  </properties>

  <!-- never deploy any poms -->
  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-deploy-plugin</artifactId>
        <version>3.0.0-M1</version>
        <configuration>
          <skip>true</skip>
        </configuration>
      </plugin>

      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-resources-plugin</artifactId>
        <version>3.1.0</version>
        <executions>
          <execution>
            <id>copy-website-includes</id>
            <phase>initialize</phase>
            <goals>
              <goal>copy-resources</goal>
            </goals>
            <configuration>
              <useDefaultDelimiters>true</useDefaultDelimiters>
              <outputDirectory>${project.build.directory}/generated-docs/website/_includes/</outputDirectory>
              <resources>
                <resource>
                  <directory>${basedir}/website/target/generated-docs/components/</directory>
                  <filtering>false</filtering>
                  <includes>
                    <include>**/*.html</include>
                    <include>**/*.css</include>
                    <include>**/*.js</include>
                    <include>**/*.jpg</include>
                    <include>**/*.jpeg</include>
                    <include>**/*.png</include>
                    <include>**/*.svg</include>
                    <include>**/*.gif</include>
                    <include>**/*.ico</include>
                  </includes>
                </resource>
              </resources>
            </configuration>
          </execution>
          <execution>
            <id>copy-css</id>
            <phase>initialize</phase>
            <goals>
              <goal>copy-resources</goal>
            </goals>
            <configuration>
              <useDefaultDelimiters>true</useDefaultDelimiters>
              <outputDirectory>${project.build.directory}/generated-docs/</outputDirectory>
              <resources>
                <resource>
                  <directory>${stylesheet.target.dir}</directory>
                  <filtering>false</filtering>
                  <includes>
                    <include>devonfw.css</include>
                  </includes>
                </resource>
              </resources>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>

</project>
```

[pom.zip](https://github.com/devonfw/docgen/files/3974075/pom.zip)

jambulud commented:
Hi, I would like to know if there is a way to conditionally add/remove the tag `<sourceDocumentName>${docgen.asciidoc.page}.${docgen.asciidoc.extension}</sourceDocumentName>` because there are use cases when you don't want to generate one single file, but multiple ones. I don't know if leaving this field empy is enough `<sourceDocumentName/>`

hohwille commented:
I am slightly confused.
```
<modules>
    <module>asciidoctor-stylesheet</module>
    <module>website</module>
    <module>search-engine</module>
  </modules>
```
What is that? This does not seem to exist here:
https://github.com/devonfw/devonfw-guide/

If you have entirely refactored `devonfw-docgen` we will end up wasting weeks and months of missunderstandings. We should be agile and do small changes to make progress from one working state to the next and feature by feature.
This feature that I provided via this PR is required for me to make further progress in `devon4j`, `devonfw-ide`, `sonar-devon-plugin`, etc. 
However, this is blocking me as we do not seem to make progress here.
All that I know is that it is somehow not working on your machine and I have no clue how to reproduce the errors you are getting. We need to find a more constructive way to get this solved together. Also I am unclear if your last question above is a new feature request that should stay completely out of the scope of this PR or if that is a feature that used to work and may now be broken as I was not aware of this feature? To me the main problem of docgen seems to be that there is no clear maintaner / product owner. There have been different people working on this with different ideas and use-cases that never really aligned their requirements.

jambulud commented:
> ```
> <modules>
>     <module>asciidoctor-stylesheet</module>
>     <module>website</module>
>     <module>search-engine</module>
>   </modules>
> ```
> 
> What is that? This does not seem to exist here:
> https://github.com/devonfw/devonfw-guide/
> 

That code comes from https://github.com/devonfw/devonfw.github.io/blob/develop/pom.xml#L15

jambulud commented:
> Hi, I would like to know if there is a way to conditionally add/remove the tag `<sourceDocumentName>${docgen.asciidoc.page}.${docgen.asciidoc.extension}</sourceDocumentName>` because there are use cases when you don't want to generate one single file, but multiple ones. I don't know if leaving this field empy is enough `<asciidoctor.sourceDocumentName/>`
> 
> https://asciidoctor.org/docs/asciidoctor-maven-plugin/#configuration

I don't need this anymore, I managed to solve it using inherit properties

https://github.com/devonfw/docgen/pull/16/files#diff-600376dffeb79835ede4a0b285078036

hohwille commented:
@jambulud I merged your additional PR #16 that is now part of this PR.
The initial "Permission denied" bug is fixed.
If I understood correctly you have already been able to build devonfw website properly with this new codebase. Is that corret? Is the "Unknown ERB implementation" error also resolved now?
Can we merge this PR and create the release? I would be happy to do so if all is ready now, but will wait for your feedback.

jambulud commented:
> If I understood correctly you have already been able to build devonfw website properly with this new codebase. Is that corret? Is the "Unknown ERB implementation" error also resolved now?

Yes, that errror is solved specifying `<eruby>erb</eruby>` in the configuration of asciidoc. That basically means "check for .erb extension when eruby templates are used"

> Can we merge this PR and create the release? I would be happy to do so if all is ready now, but will wait for your feedback.

Yes, you can do it.

hohwille commented:
@jambulud Awesome. Thanks 👍 

> Yes, that errror is solved specifying <eruby>erb</eruby> in the configuration of asciidoc.

Should we extend docgen to do that for the user? In the root POM or only in the `html` POM?

jambulud commented:
> Should we extend docgen to do that for the user? In the root POM or only in the `html` POM?

This is already being done here https://github.com/devonfw/docgen/blob/master/pom.xml#L283

But you could use a property instead of the fixed value so users can use different extensions from their pom

hohwille commented:
Awesome. Now all is clear. Will do that tiny improvement and then push out a new release.

hohwille commented:
@jambulud Thanks for all your support 👍
